rust/src/producers/circuit_generator.rs

```diff
@@ -0,0 +1,263 @@
+    let wit_idx = ((ins_nbr+1)..(ins_nbr + wit_nbr + 1)).collect::<Vec<u64>>();
+
+    let mut constraints_vec: Vec<((Vec<u64>, Vec<u8>), (Vec<u64>, Vec<u8>), (Vec<u64>, Vec<u8>))> = vec![
+        // (A ids values)  *  (B ids values)  =  (C ids values)
+        ((vec![0], vec![1]), (wit_idx.clone(), vec![1; wit_idx.len()]), (wit_idx.clone(), vec![1; wit_idx.len()])),
+    ];
+
+    for _i in 0..ins_nbr {
+        let b1: Vec<u8> = (0..wit_nbr).map(|_| rng.gen_range(0, 2)).collect();
+        let b2: Vec<u8> = (0..wit_nbr).map(|_| rng.gen_range(0, 2)).collect();
+
+        let mut result_of_equation = compute_equation(&witnesses, &b1) * compute_equation(&witnesses, &b2);
+        result_of_equation %= &builder.sink.modulus;
+        let buf = serialize_biguint(result_of_equation, size_in_bytes);
+
+        let instance_id = builder.allocate_instance_var(&buf);
+
+        let mut new_constraint: Vec<((Vec<u64>, Vec<u8>), (Vec<u64>, Vec<u8>), (Vec<u64>, Vec<u8>))> = vec![
+            ((wit_idx.clone(), b1), (wit_idx.clone(), b2), (vec![instance_id], vec![1]))
+        ];
+        constraints_vec.append(&mut new_constraint);
+        if constraints_vec.len() >= 100000 {
+            builder.push_constraints(ConstraintSystem::from(constraints_vec.as_ref()))?;
+            constraints_vec = empty_constraints.clone();
+        }
+
+    }
+
+    // if it remains constraints not yet pushed, push them.
+    if constraints_vec.len() > 0 {
+        builder.push_constraints(ConstraintSystem::from(constraints_vec.as_ref()))?;
+    }
+
+    let witness_buffer = serialize_biguints(witnesses, size_in_bytes);
+
+    builder.push_witness(Witness {
+        assigned_variables: Variables {
+            variable_ids: wit_idx, // xx, yy
+            values: Some(witness_buffer),
+        }
+    })?;
+    builder.header.free_variable_id += wit_nbr + 1;
+    builder.finish_header()?;
+    Ok(())
+}
+
+/// serialize `BigUint` as a vector of u8 (in Little-endian), of a given size, padded with 0's if required.
+fn serialize_biguint(biguint: BigUint, byte_len: usize) -> Vec<u8> {
+    let mut ret: Vec<u8> = Vec::new();
+    let mut binary = biguint.to_bytes_le();
+    let len = binary.len();
+    ret.append(&mut binary);
+    ret.append(&mut vec![0; byte_len - len]);
+
+    ret
+}
+
+/// serialize a `Vec<BigUint>` as a vector of u8, each `BigUint` exported a fixed-size vector of u8
+/// padded with 0's if required.
+fn serialize_biguints(biguints: Vec<BigUint>, byte_len: usize) -> Vec<u8> {
+    let mut ret: Vec<u8> = Vec::new();
+
+    for biguint in biguints.iter() {
+        ret.append(&mut serialize_biguint(biguint.clone(), byte_len));
+    }
+    ret
+}
+
+fn compute_equation(witnesses: &Vec<BigUint>, bit_vector: &[u8]) -> BigUint {
+    let mut ret = BigUint::zero();
+    let _: Vec<usize> = witnesses.iter().enumerate()
+                .filter(|(idx, _buint)| bit_vector[*idx] != 0 )
```

Comment:
`_buint` is unused, and the filter is not very intuetive. 
I would go with somthing like
```rust
fn compute_equation(witnesses: &Vec<BigUint>, bit_vector: &Vec<u8>) -> BigUint {
    witnesses.iter().enumerate().map(|(idx, buint)| {
        buint * bit_vector[idx];
    }).sum::<BigUint>();
}
```

Reply:
I finally chose this even cleaner solution:
`
witnesses.iter().zip(bit_vector.iter()).map(|(idx, buint)| buint * idx ).sum::<BigUint>()
`
which ensures that only slices of same length will be used (avoiding reading out of bounds).